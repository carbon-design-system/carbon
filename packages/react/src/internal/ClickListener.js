/**
 * Copyright IBM Corp. 2016, 2018
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import PropTypes from 'prop-types';
import React from 'react';

/**
 * Generic component used for reacting to a click event happening outside of a
 * given `children` element.
 */
export default class ClickListener extends React.Component {
  static propTypes = {
    children: PropTypes.element.isRequired,
    onClickOutside: PropTypes.func.isRequired,
  };

  static getEventTarget(evt) {
    // support Shadow DOM
    if (evt.composed && typeof evt.composedPath === 'function') {
      return evt.composedPath()[0];
    }
    return evt.target;
  }

  constructor(props) {
    super(props);
    // We manually bind handlers in this Component, versus using class
    // properties, so that we can properly test the `handleRef` handler with
    // enzyme.
    this.handleRef = this.handleRef.bind(this);
    this.handleDocumentClick = this.handleDocumentClick.bind(this);
  }

  componentDidMount() {
    document.addEventListener('click', this.handleDocumentClick);
  }

  componentWillUnmount() {
    document.removeEventListener('click', this.handleDocumentClick);
  }

  handleDocumentClick(evt) {
    if (this.element) {
      if (
        this.element.contains &&
        !this.element.contains(ClickListener.getEventTarget(evt))
      ) {
        this.props.onClickOutside(evt);
      }
    }
  }

  handleRef(el) {
    const { children } = this.props;
    this.element = el;

    /**
     * One important note, `children.ref` corresponds to a `ref` prop passed in
     * directly to the child, not necessarily a `ref` defined in the component.
     * This means that here we target the following `ref` location:
     *
     * <ClickListener onClickOutside={() => {}}>
     *   <Child ref={targetedRefHere} />
     * </ClickListener>
     */
    if (children.ref && typeof children.ref === 'function') {
      children.ref(el);
    }
  }

  render() {
    return React.cloneElement(this.props.children, {
      ref: this.handleRef,
    });
  }
}
