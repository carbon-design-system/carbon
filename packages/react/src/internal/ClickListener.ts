/**
 * Copyright IBM Corp. 2016, 2025
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React, {
  cloneElement,
  useEffect,
  useRef,
  type ReactElement,
} from 'react';

interface ClickListenerProps {
  children: ReactElement;
  onClickOutside: (event: MouseEvent) => void;
}

export const ClickListener = ({
  children,
  onClickOutside,
}: ClickListenerProps) => {
  const elementRef = useRef<HTMLElement | null>(null);

  const getEventTarget = (event: MouseEvent) => {
    if (event.composed && typeof event.composedPath === 'function') {
      return event.composedPath()[0];
    }

    return event.target;
  };

  const handleDocumentClick = (event: MouseEvent) => {
    if (elementRef.current?.contains) {
      const eventTarget = getEventTarget(event);

      if (
        eventTarget instanceof Node &&
        !elementRef.current.contains(eventTarget)
      ) {
        onClickOutside(event);
      }
    }
  };

  useEffect(() => {
    document.addEventListener('click', handleDocumentClick);

    return () => {
      document.removeEventListener('click', handleDocumentClick);
    };
  }, [onClickOutside]);

  const handleRef = (el: HTMLElement | null) => {
    elementRef.current = el;

    if ('ref' in children && typeof children.ref === 'function') {
      children.ref(el);
    }
  };

  return cloneElement(children, { ref: handleRef });
};
