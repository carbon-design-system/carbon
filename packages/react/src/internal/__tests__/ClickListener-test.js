/**
 * Copyright IBM Corp. 2016, 2018
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';
import ClickListener from '../ClickListener';
import { shallow, mount } from 'enzyme';

describe('ClickListener', () => {
  let onClickOutside;
  let handleRefSpy;

  beforeEach(() => {
    onClickOutside = jest.fn();
    handleRefSpy = jest.spyOn(ClickListener.prototype, 'handleRef');
  });

  afterEach(() => {
    handleRefSpy.mockRestore();
  });

  it('should render', () => {
    const wrapper = shallow(
      <ClickListener onClickOutside={onClickOutside}>
        <div>
          <div className="child">Test</div>
          <div className="child">Test</div>
        </div>
      </ClickListener>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should throw a PropType validation error if passed multiple children', () => {
    // We need to make assertions on `console.error` that come from React, so let's
    // disable the `no-console` eslint rules on certain lines.
    // eslint-disable-next-line no-console
    const originalConsoleError = console.error;
    const mockConsoleError = jest.fn();
    // eslint-disable-next-line no-console
    console.error = mockConsoleError;

    shallow(
      <ClickListener onClickOutside={onClickOutside}>
        <div className="child">Test</div>
        <div className="child">Test</div>
      </ClickListener>
    );
    expect(mockConsoleError).toHaveBeenCalledWith(
      'Warning: Failed %s type: %s%s',
      'prop',
      'Invalid prop `children` of type `array` supplied to `ClickListener`, expected a single ReactElement.',
      expect.anything() // this argument is the stack trace/report of the source file line and can change
    );

    // eslint-disable-next-line no-console
    console.error = originalConsoleError;
  });

  it('should invoke onClickOutside if click is outside of the component', () => {
    mount(
      <ClickListener onClickOutside={onClickOutside}>
        <div>
          <div className="child">Test</div>
          <div className="child">Test</div>
        </div>
      </ClickListener>
    );

    const evt = new MouseEvent('click');
    document.dispatchEvent(evt);

    expect(onClickOutside).toHaveBeenCalled();
  });

  it('should not overwrite any children function refs', () => {
    const mockRef = jest.fn();
    class Child extends React.Component {
      render() {
        return <div />;
      }
    }
    mount(
      <ClickListener onClickOutside={onClickOutside}>
        <Child ref={mockRef} />
      </ClickListener>
    );
    expect(handleRefSpy).toHaveBeenCalledTimes(1);
    expect(mockRef).toHaveBeenCalledTimes(1);
  });
});
