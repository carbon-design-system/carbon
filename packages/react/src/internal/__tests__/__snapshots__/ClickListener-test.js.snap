// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClickListener renders as expected - Component API should render 1`] = `
<div>
  <div>
    <div
      class="child"
    >
      Test
    </div>
    <div
      class="child"
    >
      Test
    </div>
  </div>
</div>
`;
